torch/_inductor/codegen/cuda/cuda_template.py

```diff
@@ -0,0 +1,179 @@
+import functools
+import itertools
+from copy import copy
+from typing import List
+from unittest.mock import patch
+
+import sympy
+
+from third_party.cutlass.tools.library import scripts as cutlass_lib
+
+from ...autotune_process import CUDABenchmarkRequest, TensorMeta
+from ...ir import Buffer, IRNode, Layout
+from ...utils import IndentedBuffer, unique
+from ...virtualized import V
+from ..common import jinja2_env
+
+from . import cutlass_utils
+from .cuda_kernel import CUDATemplateCaller, CUDATemplateKernel
+
+
+class CUDATemplate:
+    index_counter = itertools.count()
+    all_templates = dict()
+
+    def __init__(self, name: str, input_nodes: List[IRNode], layout: Layout):
+        super().__init__()
+        self.name = name
+        assert name not in self.all_templates, "duplicate template name"
+        self.all_templates[name] = self
+        self.input_nodes = input_nodes
+        self.output_node = Buffer("buf_out", layout)
+
+    @staticmethod
+    def _template_from_string(source):
+        env = jinja2_env()
+        if env is not None:
+            return env.from_string(source)
+        return None
+
+    def maybe_append_choice(self, choices, **kwargs):
+        choices.append(self.generate(**kwargs))
+
+    @staticmethod
+    def _fake_get_dtype(fake_out):
+        _get_dtype_real = V.graph.get_dtype
+
+        def get_dtype(name):
+            if name == fake_out.get_name():
+                return fake_out.get_dtype()
+            return _get_dtype_real(name)
+
+        return get_dtype
+
+    def generate(self, **kwargs) -> CUDATemplateCaller:
+        kernel_name = f"cuda_{self.name}"
+        with patch.object(
+            V.graph, "get_dtype", self._fake_get_dtype(self.output_node)
+        ), CUDATemplateKernel(
+            kernel_name=kernel_name,
+        ) as kernel:
+            # need to do call render twice to get all the needed args right
+            self.render(kernel=kernel, **kwargs)
+            code = self.render(kernel=kernel, **kwargs)
```

Comment:
I actually removed this from Triton template since it confused both you and Horace, see:
https://github.com/pytorch/pytorch/pull/105987